File: src/main/java/com/github/lppedd/inspections/InspectionsBundle.java
package com.github.lppedd.inspections;

import java.lang.ref.Reference;
import java.util.ResourceBundle;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.PropertyKey;

import com.intellij.BundleBase;
import com.intellij.reference.SoftReference;

/**
 * @author Edoardo Luppi
 */
public final class InspectionsBundle {
  private static final String BUNDLE = "messages.AdditionalInspections";
  private static Reference<ResourceBundle> bundleReference;

  public static @NotNull String get(
      @PropertyKey(resourceBundle = BUNDLE) final @NotNull String key,
      final Object... params) {
    return BundleBase.message(getBundle(), key, params);
  }

  private static @NotNull ResourceBundle getBundle() {
    var bundle = SoftReference.dereference(bundleReference);

    if (bundle == null) {
      bundle = ResourceBundle.getBundle(BUNDLE);
      bundleReference = new java.lang.ref.SoftReference<>(bundle);
    }

    return bundle;
  }
}

File: src/main/kotlin/com/github/lppedd/inspections/InspectionsBundle.kt
package com.github.lppedd.inspections

import com.intellij.BundleBase
import com.intellij.reference.SoftReference
import org.jetbrains.annotations.PropertyKey
import java.lang.ref.Reference
import java.util.*
import java.lang.ref.SoftReference as JavaSoftReference

/**
 * @author Edoardo Luppi
 */
object InspectionsBundle {
  private const val BUNDLE = "messages.AdditionalInspections"

  private var bundleReference: Reference<ResourceBundle>? = null
  private val bundle: ResourceBundle by lazy {
    SoftReference.dereference(bundleReference)
      ?: ResourceBundle.getBundle(BUNDLE).also {
        bundleReference = JavaSoftReference(it)
      }
  }

  @JvmStatic
  operator fun get(@PropertyKey(resourceBundle = BUNDLE) key: String, vararg params: Any): String =
    BundleBase.message(bundle, key, *params)
}
